### ISSUE TITLE
Stack allows a non-concrete GHC version in resolver field

### ISSUE THREAD

**Issue by mboes** [commercialhaskell/stack#2257]
For reproducibility, Stack requires a concrete nightly, a concrete LTS, or a concrete compiler name in the `resolver:` field of `stack.yaml`, while on the command-line symbolic shortcuts like `nightly` are permitted.

.... or so I thought. But the following works on current `master`:

```
$ stack new test
$ cd test
$ echo -e "resolver: ghc-7.10\npackages: ['.']"
$ stack build
```

Likewise, `ghc-8` is permitted and implicitly resolves to `ghc-8.0.1`. Presumably to `ghc-8.0.2` in the future.

**Comment by mgsloan** [comment]
> Likewise, `ghc-8` is permitted and implicitly resolves to `ghc-8.0.1`. Presumably to `ghc-8.0.2` in the future.

See the docs for `compiler-match` for more info: http://docs.haskellstack.org/en/stable/yaml_configuration/#compiler-check

With `compiler-match: match-minor`, the default, it will indeed allow `ghc-8.0.2`, but not `ghc-8.1.0`.  The reasoning is:

1) GHC developers are responsible and hopefully didn't introduce substantial breakage
2) This makes it easier to use stack with a custom built ghc / somewhat newer ghc.

Yes, it is a weakening of build determinism.  That's what `compiler-match: match-exact` is for, if you need it.

**Comment by mgsloan** [comment]
I've opened up https://github.com/commercialhaskell/stack/issues/2260 about making the version specification of the compiler resolver more consistent with snapshot resolver.
